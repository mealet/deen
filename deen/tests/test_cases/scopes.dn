struct Scope {
  scope: *char,

  fn enter(name: *char) Scope {
    println!("Entered `{}` scope", name);
    return Scope { .scope = name };
  }

  fn drop(&self) {
    println!("Exited `{}` scope", self.scope);
  }
}

fn main() i32 {
  // MAIN SCOPE
  let main_scope = Scope.enter("MAIN");

  {
    // TEMP SCOPE
    let temp_scope = Scope.enter("TEMP");
    println!("message from temp");
  }

  if 1 == 1 {
    // IF SCOPE
    let if_scope = Scope.enter("IF");
    println!("condition is true");
  } else {
    // ELSE SCOPE
    let else_scope = Scope.enter("ELSE");
  }

  return 0;
}
