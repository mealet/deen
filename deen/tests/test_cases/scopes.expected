Entered `MAIN` scope
Entered `TEMP` scope
message from temp
Exited `TEMP` scope
Entered `IF` scope
condition is true
Exited `IF` scope
Exited `MAIN` scope
